### PR TITLE
Implement fuse option for delayed objects

### LINKED ISSUE #11436: Make it easier to enable linear fusion for delayed objects
Delayed can be used to express embarrassingly parallel chains of operations. Certain configurations benefit from processing one chain completely before starting a new one (high memory usage for example). This is currently not possible by default because other tasks will almost certainly run before follow up tasks are scheduled. Linear fusion is the obvious solution for this which ensures that there is only a single task in the dask sense to express the whole chain. Enabling this is currently a real PITA:

```

import dask
from dask.optimization import cull, fuse
from dask.utils import ensure_dict
from dask import config, core

def fuse_delayed(dsk, keys, *args, **kwargs):
    dependencies = dsk.get_all_dependencies()
    dsk = ensure_dict(dsk)

    fuse_subgraphs = config.get("optimization.fuse.subgraphs")
    if fuse_subgraphs is None:
        fuse_subgraphs = True
    dsk, _ = fuse(
        dsk,
        keys,
        dependencies=dependencies,
        fuse_subgraphs=fuse_subgraphs,
    )
    dsk, _ = cull(dsk, keys)
    return dsk
    
dask.config.set({"delayed_optimize": fuse_delayed})
```

This is pretty ugly. Even if we ignore this, you still have to put a function into the dask config which isn't great either.

Ideally, we add a configuration like

```
dask.config.set({"enable-delayed-fusion": True})
```

Even better, we turn this on by default, but one step at a time

### PR DESCRIPTION
**Pull request by phofl**
- [ ] Closes #11436
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: dask/dask.yaml [hunk 1]
```diff
@@ -43,6 +43,9 @@ optimization:
     subgraphs: null  # true for dask.dataframe, false for everything else
     rename-keys: true
 
+delayed:
+  fuse: false
```
Comment (fjetter) [inline dask/dask.yaml:47]: mixed feelings about the config name. There is a dedicated `optimization` section and the other collections don't have any optimization controlling options.

Comment (phofl) [reply dask/dask.yaml:47]: putting it into the fuse section with

```
optimisations:
   fuse:
       delayed
```

messes with the optimiser :(

Comment (fjetter) [reply dask/dask.yaml:47]: what? how? Well... never mind then 💢

Comment (phofl) [reply dask/dask.yaml:48]: I **think** this was caused by a typo on my end :(

Changed things to put delayed under fuse

**Comment by rjzamora** [comment]
Seems reasonable. Just curious: Will this optimization only be applied when you directly compute/persist delayed objects? (This wouldn't kick in if you did something like `dd.from_delayed` ?)

**Comment by phofl** [comment]
only if your result is a delayed object, so no for the dataframe case

File: dask/tests/test_delayed.py [hunk 2]
```diff
@@ -855,3 +856,27 @@ def add(x, y):
...
+        return i + 3
+
+    obj = test3(test2(test(10)))
+    dsk = dict(collections_to_dsk([obj]))
+    assert len(dsk) == 3
+
+    obj2 = test3(test2(test(10)))
+    with dask.config.set({"optimization.fuse.delayed": True}):
+        dsk2 = dict(collections_to_dsk([obj]))
+    assert len(dsk2) == 2
```
Comment (rjzamora) [inline dask/tests/test_delayed.py:881]: Why is this 2 rather than 1?

File: dask/tests/test_delayed.py [hunk 3]
```diff
@@ -855,3 +856,27 @@ def add(x, y):
...
+    assert dask.compute(obj) == dask.compute(obj2)
```
Comment (rjzamora) [inline dask/tests/test_delayed.py:882]: Curious what you are testing here? (both `obj` and `obj2`) are defined as `test3(test2(test(10)))`.

Comment (phofl) [reply dask/tests/test_delayed.py:882]: good point, I intended to compute the first on in the with statement, not sure why I didn't do that

File: dask/tests/test_delayed.py (hunk 2 above)
Comment (rjzamora) [reply dask/tests/test_delayed.py:881]: Okay, that makes sense.

File: dask/tests/test_delayed.py (hunk 3 above)
Comment (phofl) [reply dask/tests/test_delayed.py]: done

File: dask/tests/test_delayed.py [hunk 4]
```diff
@@ -855,3 +856,28 @@ def add(x, y):
...
+
+    @delayed
+    def test3(i):
+        return i + 3
+
+    obj = test3(test2(test(10)))
+    dsk = dict(collections_to_dsk([obj]))
+    assert len(dsk) == 3
+
+    obj2 = test3(test2(test(10)))
```
Comment (rjzamora) [inline dask/tests/test_delayed.py:878]: Non-blocking: I don't think I understand why `obj2` is needed for this test.

**Review by rjzamora — APPROVED** [review]

File: dask/tests/test_delayed.py (hunk 4 above)
Comment (phofl) [reply dask/tests/test_delayed.py:878]: I'd rather be overly careful even though it shouldn't matter. Calling anything on a delayed object could have side-effects (even though it shouldn't here)

**Pull request merged**